pkg/server/openapiv3/servicecache.go

```diff
@@ -0,0 +1,298 @@
+	"k8s.io/kube-openapi/pkg/spec3"
+	"k8s.io/utils/lru"
+)
+
+const (
+	// DefaultServiceCacheSize is the default size of the OpenAPI service cache.
+	// Equal API configurations in multiple workspaces are shared.
+	DefaultServiceCacheSize = 100
+
+	// TODO(sttts): move to central place in kube.
+	BoundAnnotationKey = "apis.kcp.io/bound-crd"
+)
+
+// WithOpenAPIv3 returns a handler that serves OpenAPI v3 specs for CRDs, not
+// forwarding /openapi/v3 requests to the delegate handler.
+func WithOpenAPIv3(handler http.Handler, c *ServiceCache) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		if r.URL.Path == "/openapi/v3" || strings.HasPrefix(r.URL.Path, "/openapi/v3/") {
+			c.ServeHTTP(w, r)
+			return
+		}
+
+		handler.ServeHTTP(w, r)
+	})
+}
+
+// ServiceCache implements a cluster-aware OpenAPI v3 handler, sharing the
+// OpenAPI service for equal API surface configurations.
+type ServiceCache struct {
+	config *common.Config
+
+	specGetter CRDSpecGetter
+	crdLister  kcp.ClusterAwareCRDClusterLister
+
+	services    *lru.Cache
+	staticSpecs map[string]cached.Data[*spec3.OpenAPI]
+}
+
+func NewServiceCache(config *common.Config, crdLister kcp.ClusterAwareCRDClusterLister, specGetter CRDSpecGetter, serviceCacheSize int) *ServiceCache {
+	return &ServiceCache{
+		config:      config,
+		specGetter:  specGetter,
+		crdLister:   crdLister,
+		services:    lru.New(serviceCacheSize),
+		staticSpecs: map[string]cached.Data[*spec3.OpenAPI]{},
+	}
+}
+
+func (c *ServiceCache) RegisterStaticAPIs(cont *restful.Container) error {
+	// create static specs
+	byGVPath := make(map[string][]*restful.WebService)
+	for _, t := range cont.RegisteredWebServices() {
+		// Strip the "/" prefix from the name
+		gvPath := t.RootPath()[1:]
+		byGVPath[gvPath] = []*restful.WebService{t}
+	}
+	for gvPath, ws := range byGVPath {
+		spec, err := builder3.BuildOpenAPISpecFromRoutes(restfuladapter.AdaptWebServices(ws), c.config)
+		if err != nil {
+			return fmt.Errorf("failed to build OpenAPI v3 spec for %s: %w", gvPath, err)
+		}
+		etag, err := computeEtag(spec)
+		if err != nil {
+			return fmt.Errorf("failed to compute OpenAPI v3 spec etag for %s: %w", gvPath, err)
+		}
+		c.staticSpecs[gvPath] = cached.NewResultOK(spec, etag)
+	}
+
+	return nil
+}
+
+func (c *ServiceCache) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	ctx := r.Context()
+
+	clusterName, err := request.ClusterNameFrom(r.Context())
+	if err != nil {
+		http.NotFound(w, r)
+		return
+	}
+
+	log := klog.FromContext(ctx).WithValues("cluster", clusterName, "path", r.URL.Path)
+
+	// get both real CRDs and bound CRD from APIBindings
+	crds, err := c.crdLister.Cluster(clusterName).List(ctx, labels.Everything())
+	if err != nil {
+		responsewriters.InternalError(w, r, err)
+		return
+	}
+
+	// operate on sorted lists to have deterministic API configuration key
+	orderedCRDs := make([]*apiextensionsv1.CustomResourceDefinition, len(crds))
```

Comment:
Would this make sense to move into apiConfiguration? 
Now `apiConfiguration` expects a sorted list. If at any point in the probable future, somebody want to reuse this function, this is not explicitly clear (function comment might be good too)?

Reply:
Ah, line 142 would need to go in. Code becomes offuscated. I think explicit code comment on function would be enough.

Reply:
have renamed the argument to orderedCRDs. Should make it clearer.